e2e: node: add feature to annotate pods with owning code

ffromani opened the pull request:
/kind cleanup

#### What this PR does / why we need it:
e2e_node test depend on very specific shared state (node state).
Pod leakages between tests oftentimes cause the test preconditions  to be silently corrupted, causing hard to debug CI failures.

We add the option to add an annotation to pods which records the code line (source code:line) which triggered the pod creation, so it becomes easier to track which test needs better cleanup.

The relevant e2e framework code is used in all the e2e suites, so to minimize any unwanted consequences we make the feature opt-in, planning to enable it initially (and likely only)  in the e2e_node tests.

For more context, please check the conversation in https://github.com/kubernetes/kubernetes/pull/123468

#### Which issue(s) this PR fixes:
Fixes #N/A

#### Special notes for your reviewer:
N/A

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

ffromani committed:
e2e: node: add code  to track the pod creator code

e2e_node test depend on very specific shared state (node state).
Pod leakages between tests oftentimes cause the test preconditions
to be silently corrupted, causing hard to debug CI failures.

We add the option to add an annotation to pods which records
the code line (source code:line) which triggered the pod creation,
so it becomes easier to track which test needs better cleanup.

The relevant e2e framework code is used in all the e2e suites,
so to minimize any unwanted consequences we make the feature
opt-in, planning to enable it initially (and likely only)
in the e2e_node tests.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani commented:
/sig node
/sig testing

ffromani commented:
/cc @pohly

pohly commented on test/e2e/framework/pod/pod_client.go:
```diff
@@ -89,43 +90,84 @@ func PodClientNS(f *framework.Framework, namespace string) *PodClient {
+	//    +- createWithOwner       # 1
+	//       +- setOwnerAnnotation # 0
+	//          +- runtime.Caller  # see https://pkg.go.dev/runtime#Caller
+	return c.createWithOwner(ctx, pod, 3)
```
There's no guarantee that `Create` gets called directly by test code. It might get called by another helper.

A more reliable way to determine the caller is to invoke `ginkgo.GinkgoHelper` in all helper functions, like this one here.

Then call [NewCodeLocation](https://pkg.go.dev/github.com/onsi/ginkgo/v2/types#NewCodeLocation) with zero skip value to get the caller.

ffromani replied:
TIL, thanks! Changing accordingly

pohly commented on test/e2e/framework/pod/pod_client.go:
```diff
@@ -192,6 +207,19 @@ func (c *PodClient) DeleteSync(ctx context.Context, name string, options metav1.
+	ginkgo.GinkgoHelper()
+	if !c.trackOwner {
+		return
+	}
```
Nit: you can return before the `GinkgoHelper` call.

ffromani replied:
thanks, will do

pohly commented on test/e2e/framework/pod/pod_client.go:
```diff
@@ -89,19 +90,32 @@ func PodClientNS(f *framework.Framework, namespace string) *PodClient {
+// which leak pods because insufficient cleanup.
+func (c *PodClient) WithOwnerTracking() *PodClient {
+	c.trackOwner = true
+	return c
```
Would it make sense to have a per-test-suite default for this? This removes the need to touch >40 files and future E2E tests will do the right thing.

This method can have an explicit `enabled bool` parameter to override that default on a per-test basis.

I also would make it so that `WithOwnerTracking` returns a shallow clone.

Otherwise this code here won't work as expected:
```
c := e2epod.NewPodClient(f)
c.WithOwnerTracking(true).Create(...) // With tracking.
c.Create(...) // Also with tracking, which is unexpected.
```

You get that implicitly by changing the receiver to a value and returning a pointer to it.

ffromani replied:
Yes, this is one the reasons why it's WIP. I'm looking for a smarter way to enable per-suite, but not obvious solutions yet, depends on how the code is laid out.

Re: shallow clone: very good idea, will do.

ffromani replied:
Went for the "global enable in the suite". The approach grew on me and the code looks better than I thought.

ffromani committed:
e2e: node: use test owner tracking

e2e_node test depend on very specific shared state (node state).
Pod leakages between tests oftentimes cause the test preconditions
to be silently corrupted, causing hard to debug CI failures.

Use the new facility to annotate pods with test owner (= the
test code which created the test) to help debug these failures.

For more context, please check the conversation in #123468

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani commented:
WIP because I want to collect example pod with annotation and record it here / in the commit message before to call it done

ffromani commented:
/retest

pohly commented:
A global default could be a variable in the test/e2e/framework/pod package. The e2e_node suite setup code then would have to change that.

ffromani commented:
> A global default could be a variable in the test/e2e/framework/pod package. The e2e_node suite setup code then would have to change that.

yup, thanks. Exactly what I did.

ffromani commented:
works as expected (example output from local tree):
```
annotations for memorymanager-metrics-2899/memmngrpodgssn4: map[string]string{"owner.test":"k8s.io/kubernetes/test/e2e_node/memory_manager_metrics_test.go:154"}
```

ffromani commented:
/retest

pohly commented on test/e2e/framework/pod/pod_client.go:
```diff
@@ -58,6 +59,11 @@ const (
 
+// global flags so we can enable features per-suite instead of per-client.
+var (
+	GlobalTrackOwner bool // see WithOwnerTracking
```
Can you add a proper godoc comment?

I don't know if people use https://pkg.go.dev/k8s.io/kubernetes/test/e2e/framework/pod, but they might.

pohly commented on test/e2e/framework/pod/pod_client.go:
```diff
@@ -89,19 +95,41 @@ func PodClientNS(f *framework.Framework, namespace string) *PodClient {
+// which leak pods because insufficient cleanup.
+// Note we want a shallow clone to avoid mutating the receiver.
+// Only PodClients returned by this helper should have tracking enabled.
+func (c PodClient) WithOwnerTracking() *PodClient {
```
I prefer a boolean parameter instead of adding WithoutOwnerTracking. That's what is done elsewhere.

pohly replied:
Also cross-reference the global default here with `[GlobalTrackOwner]`.

ffromani replied:
Sorry, missed the convention. Fixing.

pohly commented on test/e2e/framework/pod/pod_client.go:
```diff
@@ -192,6 +221,19 @@ func (c *PodClient) DeleteSync(ctx context.Context, name string, options metav1.
 
+// addTestOrigin adds annotations to help identifying tests which incorrectly leak pods because insufficient cleanup
+func (c *PodClient) setOwnerAnnotation(pod *v1.Pod) {
+	if !GlobalTrackOwner && !c.trackOwner {
```
With this check, it's impossible to set `GlobalTrackOwner=true` and then disable it in the `PodClient` instance.

Constructing the `PodClient` must initialize itself with the `GlobalTrackOwner`, then only `c.trackOwner` gets checked here.

ffromani replied:
right. Done, thanks

pohly commented on test/e2e/framework/pod/pod_client.go:
```diff
@@ -56,6 +57,14 @@ const (
+
+// global flags so we can enable features per-suite instead of per-client.
+var (
+	GlobalOwnerTracking bool // see WithOwnerTracking
```
godoc comment?

ffromani replied:
sorry, fixed.

pohly reviewed (commented):
Looks good to me from a test/e2e/framework perspective. I'll let some other SIG Node developer do the LGTM.

/approve

haircommander commented:
great idea!

/lgtm

ffromani commented:
/hold

ffromani commented:
running  some serial suite just in case, better a bit extra safe than sorry

ffromani commented:
/test

ffromani commented:
/test pull-kubernetes-e2e-capz-windows-serial-slow
/test pull-kubernetes-node-e2e-containerd-serial-ec2
/test pull-kubernetes-node-kubelet-serial-containerd

ffromani commented:
swap failures: https://github.com/kubernetes/kubernetes/pull/125402

ffromani commented:
/retest-required

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager
/test pull-kubernetes-node-swap-conformance-fedora-serial
/test pull-kubernetes-node-swap-conformance-ubuntu-serial

ffromani commented:
/hold cancel

all the failures are known and there is no obvious link with this PR
